lib/debug/session.rb

```diff
@@ -996,8 +996,13 @@ class << self
       # ::DEBUGGER__.add_catch_breakpoint 'RuntimeError'
 
       Binding.module_eval do
-        ::DEBUGGER__.add_line_breakpoint __FILE__, __LINE__ + 1
-        def bp; nil; end
+        def bp(commands: nil)
+          if commands
+            cmds = commands.split(";;")
+            ::DEBUGGER__::SESSION.add_initial_commands cmds
+          end
+          ::DEBUGGER__.add_line_breakpoint __FILE__, __LINE__ + 1
+        end
```

Comment:
why move it?

Reply:
the original breakpoint points to the start of the method:

```
❯ exe/rdbg target.rb
[1, 10] in target.rb
=>    1| class Foo
      2|   def first_call
      3|     second_call(20)
      4|   end
      5|
      6|   def second_call(num)
      7|     third_call_with_block do |ten|
      8|       forth_call(num, ten)
      9|     end
     10|   end
=>#0    <main> at target.rb:1

(rdbg) c
[14, 23] in target.rb
     14|
     15|     yield(10)
     16|   end
     17|
     18|   def forth_call(num1, num2)
=>   19|     binding.bp(commands: "bt ;; info")
     20|     num1 + num2
     21|   end
     22| end
     23|
=>#0    Foo#forth_call(num1=20, num2=10) at target.rb:19
  #1    block {|ten=10|} in second_call at target.rb:8
  # and 4 frames (use `bt' command for all frames)

Stop by #0 line bp /Users/st0012/projects/debug/lib/debug/session.rb:1000 (call)

(rdbg) c
```

```
        # line 1000
        def bp(commands: nil)
```

so it's activated before the commands are injected in the method body.

by moving the breakpoint here, we can be sure that it's triggered after the injection is completed.

Reply:
I see. But it increases bps when calling `bp` method (see bps for `break` commadn).
I remember that this is why I make an empty method.

I close this issue and I try to implement it.

Reply:
I see. can you later link the commit here so I can learn the implementation as well? thanks.

Reply:
https://github.com/ruby/debug/commit/3f4ded69301d3e9552874d01e0a41b1a01032e42